mountpoint-s3/scripts/fs_bench.sh

```diff
@@ -80,15 +84,118 @@ run_fio_job() {
       else $job.write.bw / 1024 end)) | {name: .name, value: (.value / .len), unit: "MiB/s"}' ${results_dir}/${job_name}_iter*.json | tee ${results_dir}/${job_name}_parsed.json
 }
 
+should_run_job() {
+    job_file=$1
+    if [[ -n "${JOB_NAME_FILTER}" ]]; then
+      if [[ "${job_file}" == *"${JOB_NAME_FILTER}"* ]]; then
+        # job name matches filter
+        return 0
+      else
+        # job name does not match filter
+        return 1
+      fi
+    fi
+}
+
+cache_benchmark () {
+  jobs_dir=mountpoint-s3/scripts/fio/read_cache
+
+  for job_file in "${jobs_dir}"/*.fio; do
+
+    if ! should_run_job "${job_file}"; then
+      echo "Skipping job ${job_file} because it does not match ${JOB_NAME_FILTER}"
+      continue
+    fi
+
+    mount_dir=$(mktemp -d /tmp/fio-XXXXXXXXXXXX)
+    # creates a cache directoy with the suffix of the mount directory
+    cache_dir=$(mktemp -d -t `basename "${mount_dir}"`-cache-XXXXXXXXXXXX)
+
+    job_name=$(basename "${job_file}")
+    job_name="${job_name%.*}"
+    log_dir=logs/${job_name}
+
+    # cleanup mount directory and log directory
+    cleanup() {
+      echo "cache_benchmark:cleanup"
+      # unmount file system only if it is mounted
+      ! mountpoint -q ${mount_dir} || sudo umount ${mount_dir}
+      sudo rm -rf ${mount_dir}
+      sudo rm -rf ${cache_dir} # umount should clean this, but just in case
+      sudo rm -rf ${log_dir}
+    }
+
+    # trap cleanup on exit
+    trap 'cleanup' EXIT
+
+    rm -rf ${log_dir}
+    mkdir -p ${log_dir}
+
+    # mount file system
+    set +e
+    cargo run --quiet --release -- \
+      ${S3_BUCKET_NAME} ${mount_dir} \
+      --debug \
+      --cache=${cache_dir} \
+      --log-directory=${log_dir} \
+      --prefix=${S3_BUCKET_TEST_PREFIX}
+    mount_status=$?
+    set -e
+    if [ $mount_status -ne 0 ]; then
+      echo "Failed to mount file system"
+      exit 1
+    fi
+
+    # set bench file
+    bench_file=${S3_BUCKET_BENCH_FILE}
+    # run against small file if the job file ends with small.fio
+    if [[ $job_file == *small.fio ]]; then
+      bench_file=${S3_BUCKET_SMALL_BENCH_FILE}
+    fi
+
+    # let's read the file so that it is cached
+    cat ${mount_dir}/${bench_file} > /dev/null
+
+    if [ -z "$(ls -A ${cache_dir}/mountpoint-cache/)" ]; then
+      echo "cache directory is empty, something is wrong..."
+      exit 1
+    fi
+
+    # run the benchmark
+    run_fio_job $job_file $bench_file $mount_dir $log_dir
```

Comment:
Also, for the cache should we take the average of 2nd to 10th iteration(rather than 1st to 10th as in case of read benchmark) of fio job as the 1st iteration would not be making any use of cache?